File: Java/DictionaryApplication.java
import java.util.Scanner;

class Student {
    String username;
    String fullName;
    int age;
    String city;
    String zipCode;
    String[] courses;
    int numberOfCourses;

    Student(String username, String fullName, int age, String city, String zipCode, String[] initialCourses) {
        this.username = username;
        this.fullName = fullName;
        this.age = age;
        this.city = city;
        this.zipCode = zipCode;
        this.courses = new String[20];
        this.numberOfCourses = 0;
        for (int courseIndex = 0; courseIndex < initialCourses.length; courseIndex++) {
            this.courses[this.numberOfCourses] = initialCourses[courseIndex];
            this.numberOfCourses++;
        }
    }
}

public class DictionaryApplication {
    static Student[] students = new Student[100];
    static int totalNumberOfStudents = 0;
    static String[] offeredCourses = {
        "Math", "Physics", "Computer Science", "Biology", "Chemistry",
        "Statistics", "English", "Economics", "History", "Philosophy",
        "Sociology", "Political Science", "Geography", "Psychology", "Art",
        "Music", "Engineering", "Law", "Medicine", "Business"
    };

    public static void createStudent(String username, String fullName, int age, String city, String zipCode, String[] initialCourses) {
        students[totalNumberOfStudents] = new Student(username, fullName, age, city, zipCode, initialCourses);
        totalNumberOfStudents++;
    }

    public static Student findStudent(String username) {
        for (int studentIndex = 0; studentIndex < totalNumberOfStudents; studentIndex++) {
            if (students[studentIndex].username.equals(username)) {
                return students[studentIndex];
            }
        }
        return null;
    }

    public static void addCourse(String username, String courseName) {
        Student student = findStudent(username);
        if (student != null && courseIsOffered(courseName)) {
            for (int courseIndex = 0; courseIndex < student.numberOfCourses; courseIndex++) {
                if (student.courses[courseIndex].equals(courseName)) return;
            }
            student.courses[student.numberOfCourses] = courseName;
            student.numberOfCourses++;
        }
    }

    public static void removeCourse(String username, String courseName) {
        Student student = findStudent(username);
        if (student != null) {
            int coursePosition = -1;
            for (int courseIndex = 0; courseIndex < student.numberOfCourses; courseIndex++) {
                if (student.courses[courseIndex].equals(courseName)) {
                    coursePosition = courseIndex;
                    break;
                }
            }
            if (coursePosition != -1) {
                for (int courseIndex = coursePosition; courseIndex < student.numberOfCourses - 1; courseIndex++) {
                    student.courses[courseIndex] = student.courses[courseIndex + 1];
                }
                student.numberOfCourses--;
            }
        }
    }

    public static void updateStudent(String username, String fullName, int age, String city, String zipCode) {
        Student student = findStudent(username);
        if (student != null) {
            student.fullName = fullName;
            student.age = age;
            student.city = city;
            student.zipCode = zipCode;
        }
    }

    public static boolean courseIsOffered(String courseName) {
        for (int courseIndex = 0; courseIndex < offeredCourses.length; courseIndex++) {
            if (offeredCourses[courseIndex].equals(courseName)) return true;
        }
        return false;
    }

    public static String[] getCourses(String username) {
        Student student = findStudent(username);
        if (student != null) {
            String[] studentCourses = new String[student.numberOfCourses];
            for (int courseIndex = 0; courseIndex < student.numberOfCourses; courseIndex++) {
                studentCourses[courseIndex] = student.courses[courseIndex];
            }
            return studentCourses;
        }
        return new String[0];
    }

    public static String getCity(String username) {
        Student student = findStudent(username);
        if (student != null) return student.city;
        return "";
    }

    public static String getZipCode(String username) {
        Student student = findStudent(username);
        if (student != null) return student.zipCode;
        return "";
    }

    public static int getTotalNumberOfStudents() {
        return totalNumberOfStudents;
    }

    public static void displayStudent(String username) {
        Student student = findStudent(username);
        if (student != null) {
            System.out.println("Username: " + student.username);
            System.out.println("Full Name: " + student.fullName);
            System.out.println("Age: " + student.age);
            System.out.println("City: " + student.city);
            System.out.println("Zip Code: " + student.zipCode);
            System.out.print("Courses: ");
            for (int index = 0; index < student.numberOfCourses; index++) {
                System.out.print(student.courses[index] + " ");
            }
            System.out.println("\n-------------------------");
        } else {
            System.out.println("Student not found.");
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("Welcome to Dictionary Application - Student Record System");

        while (true) {
            System.out.println("\n1. Create Student\n2. Display Student\n3. Add Course\n4. Remove Course\n5. Update Student\n6. Total Students\n7. Exit");
            System.out.print("Choose option: ");
            int choice = scanner.nextInt();
            scanner.nextLine();

            if (choice == 1) {
                System.out.print("Enter username: ");
                String username = scanner.nextLine();
                System.out.print("Enter full name: ");
                String fullName = scanner.nextLine();
                System.out.print("Enter age: ");
                int age = scanner.nextInt();
                scanner.nextLine();
                System.out.print("Enter city: ");
                String city = scanner.nextLine();
                System.out.print("Enter zip code: ");
                String zipCode = scanner.nextLine();
                System.out.print("Enter number of courses: ");
                int numberOfCourses = scanner.nextInt();
                scanner.nextLine();
                String[] courses = new String[numberOfCourses];
                for (int i = 0; i < numberOfCourses; i++) {
                    System.out.print("Course " + (i+1) + ": ");
                    courses[i] = scanner.nextLine();
                }
                createStudent(username, fullName, age, city, zipCode, courses);
            } else if (choice == 2) {
                System.out.print("Enter username: ");
                String username = scanner.nextLine();
                displayStudent(username);
            } else if (choice == 3) {
                System.out.print("Enter username: ");
                String username = scanner.nextLine();
                System.out.print("Enter course to add: ");
                String courseName = scanner.nextLine();
                addCourse(username, courseName);
            } else if (choice == 4) {
                System.out.print("Enter username: ");
                String username = scanner.nextLine();
                System.out.print("Enter course to remove: ");
                String courseName = scanner.nextLine();
                removeCourse(username, courseName);
            } else if (choice == 5) {
                System.out.print("Enter username: ");
                String username = scanner.nextLine();
                System.out.print("Enter new full name: ");
                String fullName = scanner.nextLine();
                System.out.print("Enter new age: ");
                int age = scanner.nextInt();
                scanner.nextLine();
                System.out.print("Enter new city: ");			
                String city = scanner.nextLine();
                System.out.print("Enter new zip code: ");
                String zipCode = scanner.nextLine();
                updateStudent(username, fullName, age, city, zipCode);
            } else if (choice == 6) {
                System.out.println("Total students: " + getTotalNumberOfStudents());
            } else if (choice == 7) {
                System.out.println("Exiting...");
                break;
            } else {
                System.out.println("Invalid choice.");
            }
        }

        scanner.close();
    }
}

File: Python/Dictionary_Application.py
available_courses = {
    "Math", "Physics", "Computer Science", "Biology", "Chemistry",
    "Statistics", "English", "Economics", "History", "Philosophy",
    "Sociology", "Political Science", "Geography", "Psychology",
    "Art", "Music", "Engineering", "Law", "Medicine", "Business"
}

students = {}

def create_student(username, name, age, courses_list, city, zip_code):
    if username in students:
        return False
    courses = set(course for course in courses_list if course in available_courses)
    students[username] = {
        "name": name,
        "age": age,
        "courses": courses,
        "address": {"city": city, "zip": zip_code}
    }
    return students[username]

def display_student(username):
    return students.get(username)

def display_courses(username):
    student = students.get(username)
    return student["courses"] if student else None

def display_zip(username):
    student = students.get(username)
    return student["address"]["zip"] if student else None

def display_city(username):
    student = students.get(username)
    return student["address"]["city"] if student else None

def add_course(username, course):
    student = students.get(username)
    if student and course in available_courses and course not in student["courses"]:
        student["courses"].add(course)
        return True
    return False

def update_course(username, old_course, new_course):
    student = students.get(username)
    if student and old_course in student["courses"] and new_course in available_courses:
        student["courses"].remove(old_course)
        student["courses"].add(new_course)
        return True
    return False

def update_student_info(username, name=None, age=None, city=None, zip_code=None):
    student = students.get(username)
    if not student:
        return False
    if name: student["name"] = name
    if age: student["age"] = age
    if city: student["address"]["city"] = city
    if zip_code: student["address"]["zip"] = zip_code
    return True

def total_students():
    return len(students)

if __name__ == "__main__":
    while True:
        print("\n=== Student Record System ===")
        print("1. Add Student")
        print("2. View Student")
        print("3. View Courses")
        print("4. Add Course")
        print("5. Update Course")
        print("6. Update Student Info")
        print("7. View City")
        print("8. View Zip")
        print("9. Total Students")
        print("0. Exit")

        choice = input("Enter choice: ")

        if choice == "1":
            username = input("Username: ")
            name = input("Name: ")
            age = int(input("Age: "))
            courses = input("Courses (comma-separated): ").split(",")
            city = input("City: ")
            zip_code = input("Zip: ")
            if create_student(username, name, age, [course.strip() for course in courses], city, zip_code):
                print("Student added!")
            else:
                print("Username already exists!")

        elif choice == "2":
            username = input("Enter username: ")
            print(display_student(username) or "Student not found.")

        elif choice == "3":
            username = input("Enter username: ")
            print(display_courses(username) or "No courses found.")

        elif choice == "4":
            username = input("Enter username: ")
            course = input("Course to add: ")
            print("Course added!" if add_course(username, course) else "Cannot add course.")

        elif choice == "5":
            username = input("Enter username: ")
            old_course = input("Old course: ")
            new_course = input("New course: ")
            print("Course updated!" if update_course(username, old_course, new_course) else "Update failed.")

        elif choice == "6":
            username = input("Enter username: ")
            name = input("New name (leave blank to skip): ")
            age_input = input("New age (leave blank to skip): ")
            city = input("New city (leave blank to skip): ")
            zip_code = input("New zip (leave blank to skip): ")
            age = int(age_input) if age_input else None
            print("Info updated!" if update_student_info(username, name or None, age, city or None, zip_code or None) else "Update failed.")

        elif choice == "7":
            username = input("Enter username: ")
            print(display_city(username) or "City not found.")

        elif choice == "8":
            username = input("Enter username: ")
            print(display_zip(username) or "Zip not found.")

        elif choice == "9":
            print("Total students:", total_students())

        elif choice == "0":
            print("Goodbye!")
            break

        else:
            print("Invalid choice. Try again.")
            strip()
